core/store/src/trie/iterator.rs

```diff
@@ -442,27 +442,138 @@ mod tests {
     /// included and the last one is excluded.
     #[test]
-    fn test_visit_interval() {
+    fn test_visit_interval_basic() {
+        init_test_logger();
         let trie_changes = vec![(b"aa".to_vec(), Some(vec![1])), (b"abb".to_vec(), Some(vec![2]))];
         let tries = create_tries();
         let state_root =
             test_populate_trie(&tries, &Trie::EMPTY_ROOT, ShardUId::single_shard(), trie_changes);
         let trie = tries.get_trie_for_shard(ShardUId::single_shard(), state_root);
-        let path_begin: Vec<_> = NibbleSlice::new(b"aa").iter().collect();
-        let path_end: Vec<_> = NibbleSlice::new(b"abb").iter().collect();
-        let mut trie_iter = trie.iter().unwrap();
-        let items = trie_iter.visit_nodes_interval(&path_begin, &path_end).unwrap();
-        let trie_items: Vec<_> = items.into_iter().map(|item| item.key).flatten().collect();
-        assert_eq!(trie_items, vec![b"aa"]);
+
+        verify_items_in_interval(b"aa", b"abb", &trie, vec![b"aa".to_vec()]);
+    }
+
+    /// Checks that for visiting interval of trie nodes first state key is
+    /// included and the last one is excluded for degenerate trie tree
+    #[test]
+    fn test_visit_interval_degenerate() {
+        let trie_changes = vec![
+            (b"a".to_vec(), Some(vec![0])),
+            (b"aa".to_vec(), Some(vec![1])),
+            (b"aaa".to_vec(), Some(vec![2])),
+            (b"aaaa".to_vec(), Some(vec![3])),
+        ];
+        let tries = create_tries();
+        let state_root =
+            test_populate_trie(&tries, &Trie::EMPTY_ROOT, ShardUId::single_shard(), trie_changes);
+        let trie = tries.get_trie_for_shard(ShardUId::single_shard(), state_root);
+
+        verify_items_in_interval(b"a", b"aa", &trie, vec![b"a".to_vec()]);
+    }
+
+    /// for boundary at leaf node
+    /// Checks that for visiting interval of trie nodes first state key is
+    /// included and the last one is excluded for boundary is leaf node
+    #[test]
+    fn test_visit_interval_boundary_at_leaf_node() {
+        init_test_logger();
+        let trie_changes = vec![
+            (b"aa".to_vec(), Some(vec![0])),
+            (b"abb".to_vec(), Some(vec![1])),
+            (b"abc".to_vec(), Some(vec![1])),
+            (b"b".to_vec(), Some(vec![2])),
+            (b"baab".to_vec(), Some(vec![3])),
+            (b"baac".to_vec(), Some(vec![4])),
+            (b"bbb".to_vec(), Some(vec![5])),
+        ];
+        let tries = create_tries();
+        let state_root =
+            test_populate_trie(&tries, &Trie::EMPTY_ROOT, ShardUId::single_shard(), trie_changes);
+        let trie = tries.get_trie_for_shard(ShardUId::single_shard(), state_root);
+
+        // begin and end both at leaf nodes
+        verify_items_in_interval(
+            b"abb",
+            b"baac",
+            &trie,
+            vec![b"abb".to_vec(), b"abc".to_vec(), b"b".to_vec(), b"baab".to_vec()],
```

Comment:
Could be nice to have a test where left and/or right boundaries don't exist as keys as well.